### ISSUE TITLE
 Failed to decode server's CA: 

### ISSUE THREAD

**Issue by arpitaso** [mostafa/xk6-kafka#222]
Hi, 
when trying to run the k6 script. I get the below error.

ERRO[0000] GoError: Failed to decode server's CA: /Users/arpita.sood/Documents/k6-code/src/resources/srkeystore.jks, OriginalError: %!w(*errors.errorString=&{entry not found})
        at github.com/mostafa/xk6-kafka.(*Kafka).loadJKSFunction-fm (native)
        at file:///Users/arpita.sood/Documents/k6-code/src/scripts/dev-test-script.js:17:20(25)  hint="script exception"

**Comment by mostafa** [comment]
Hey @arpitaso,

Can you give me a sample of the script, specifically the TLS configs. Please remove sensitive information before posting.

**Comment by arpitaso** [comment]
hi @mostafa 


/*
 * This script shows how to load certificates and keys from
 * a JKS keystore, so that they can be used in a configuring TLS.
 * This is just a showcase, and not a complete script.
 * The keystore MUST be created with JKS storetype.
 *
 * ⚠️ The PKCS#12 format is not supported.
 */

import { LoadJKS, TLS_1_2 } from "k6/x/kafka";

// If server and client keystore are separate, then you must
// call LoadJKS twice, once for each keystore.
// This will load the certificates and keys from the keystore
// and write them to the disk, so that they can be used in
// the TLS configuration.
const jks = LoadJKS({
  path: "./kaflafiles/kafka-keystore.jks",
  password: "--------"
});
const tlsConfig = {
  enableTls: true,
  insecureSkipTlsVerify: false,
  minVersion: TLS_1_2,

  // The certificates and keys can be loaded from a JKS keystore:
  // clientCertsPem is an array of PEM-encoded certificates, and the filenames
  // will be named "client-cert-0.pem", "client-cert-1.pem", etc.
  // clientKeyPem is the PEM-encoded private key and the filename will be
  // named "client-key.pem".
  // serverCaPem is the PEM-encoded CA certificate and the filename will be
  // named "server-ca.pem".
  clientCertPem: jks["clientCertsPem"][0], // The first certificate in the chain
  clientKeyPem: jks["clientKeyPem"],
  serverCaPem: jks["serverCaPem"],
};

export default function () {
  console.log(Object.keys(jks));
  console.log(jks);
  console.log(tlsConfig);
}

**Comment by mostafa** [comment]
@arpitaso 

The key aliases are important, and must be set, otherwise it will be an empty string and won't match anything inside the keystore you passed:

```javascript
const jks = LoadJKS({
  path: "fixtures/kafka-keystore.jks",
  password: "password",
  clientCertAlias: "localhost",
  clientKeyAlias: "localhost",
  clientKeyPassword: "password",
  serverCaAlias: "caroot",
});
```

If you are not sure about the key aliases, I recommend using the [keystore explorer](https://keystore-explorer.org/) tool to open your JKS file and see the filename, aka. aliases.

**Comment by arpitaso** [comment]
@mostafa 

 clientCertAlias: "localhost",
 clientKeyAlias: "localhost",
 clientKeyPassword: "password",
 serverCaAlias: "caroot",
  
 1. can you please elaborate the meaning of the above keys, so that i can pass the desired values to the above keys.
 2. do i have to check the keystore.jks file only to enter the above values?

**Comment by arpitaso** [comment]
@mostafa 
my keystore file looks like this
and my trustore file looks like this.

i see that the keystore.jks does not have a "trusted certificate entry" and the "trusted certificate entry" is available in truststore.jks.
PFA attached ss of how both the files look 

do i have to explicitly add "trusted certificate entry" in keystore.jks to establish ssl connection?

<img width="554" alt="keystore" src="https://github.com/mostafa/xk6-kafka/assets/133948820/628bc366-8963-4e86-bfb7-8c8e338e553d">
<img width="581" alt="truststore" src="https://github.com/mostafa/xk6-kafka/assets/133948820/5edac50b-7728-4561-890b-eb4a0a86b9ca">

**Comment by mostafa** [comment]
The `Entry Name` of the certificate and key is the alias I am referring to. If the server certificate is in a separate file, you have to use `LoadJKS` twice with different files to extract both.

**Comment by arpitaso** [comment]
@mostafa 

Below is my js script

```
/*
 * This script shows how to load certificates and keys from
 * a JKS keystore, so that they can be used in a configuring TLS.
 * This is just a showcase, and not a complete script.
 * The keystore MUST be created with JKS storetype.
 *
 * ⚠️ The PKCS#12 format is not supported.
 */

import { LoadJKS, TLS_1_2 } from "k6/x/kafka";

// If server and client keystore are separate, then you must
// call LoadJKS twice, once for each keystore.
// This will load the certificates and keys from the keystore
// and write them to the disk, so that they can be used in
// the TLS configuration.
const jks = LoadJKS({
  path: "/Users/arpita.sood/Documents/k6-code/src/resources/srkeystore.jks",
  password: "nxs9a08DskoOpas2d34lwEs",
  clientCertAlias: "retina-kafka",
  clientKeyAlias: "retina-kafka",
  clientKeyPassword: "nxs9a08DskoOpas2d34lwEs",
  serverCaAlias: "rootca",
});

 jks = LoadJKS({
  path: "/Users/arpita.sood/Documents/k6-code/src/resources/truststore.jks",
  password: "nxs9a08DskoOpas2d34lwEs",
  clientCertAlias: "retina-kafka",
  clientKeyAlias: "retina-kafka",
  clientKeyPassword: "nxs9a08DskoOpas2d34lwEs",
  serverCaAlias: "rootca",
});

//const jks1 = LoadJKS({
//  path: "/Users/arpita.sood/Documents/k6-code/src/resources/truststore.jks",
//  password: "nxs9a08DskoOpas2d34lwEs",
////  clientCertAlias: "retina-kafka",
////  clientKeyAlias: "retina-kafka",
////  clientKeyPassword: "nxs9a08DskoOpas2d34lwEs",
//  serverCaAlias: "rootca",
//});



const tlsConfig = {
  enableTls: true,
  insecureSkipTlsVerify: false,
  minVersion: TLS_1_2,

  // The certificates and keys can be loaded from a JKS keystore:
  // clientCertsPem is an array of PEM-encoded certificates, and the filenames
  // will be named "client-cert-0.pem", "client-cert-1.pem", etc.
  // clientKeyPem is the PEM-encoded private key and the filename will be
  // named "client-key.pem".
  // serverCaPem is the PEM-encoded CA certificate and the filename will be
  // named "server-ca.pem".
  clientCertPem: jks["clientCertsPem"][0], // The first certificate in the chain
  clientKeyPem: jks["clientKeyPem"],
  serverCaPem: jks["serverCaPem"],

//
//    clientCertPem: jks1["clientCertsPem"][0], // The first certificate in the chain
//    clientKeyPem: jks1["clientKeyPem"],
//    serverCaPem: jks1["serverCaPem"],
};

export default function () {
  console.log(Object.keys(jks));
  console.log(jks);
  console.log(tlsConfig);
}
```

I followed what you said but i still get the below error
```
ERRO[0000] GoError: Failed to decode server's CA: /Users/arpita.sood/Documents/k6-code/src/resources/srkeystore.jks, OriginalError: %!w(*errors.errorString=&{entry not found})
        at github.com/mostafa/xk6-kafka.(*Kafka).loadJKSFunction-fm (native)
        at file:///Users/arpita.sood/Documents/k6-code/src/scripts/dev-test-script.js:17:20(33)  hint="script exception"
```

**Comment by arpitaso** [comment]
@mostafa  can you please help here.?
Thanks

**Comment by mostafa** [comment]
@arpitaso 

You should create two separate objects when calling `LoadJKS` on multiple files. Don't reuse `jks` constant.

**Comment by luciowu-airwallex** [comment]
@mostafa , Hello Mostafa,
I met a similar error 
```
import { LoadJKS, TLS_1_2 } from "k6/x/kafka";

// If server and client keystore are separate, then you must
// call LoadJKS twice, once for each keystore.
// This will load the certificates and keys from the keystore
// and write them to the disk, so that they can be used in
// the TLS configuration.
const jks = LoadJKS({
    path: "../kafka/truststore.jks",
    password: "1234567",
    serverCaAlias: "kafka",
});
const tlsConfig = {
    enableTls: true,
    insecureSkipTlsVerify: false,
    minVersion: TLS_1_2,


};

export default function () {
    console.log(Object.keys(jks));
    console.log(jks);
    console.log(tlsConfig);
}
```
But we still get the exception
_ERRO[0000] GoError: Failed to decode client's private key: ***/user/workspace_k6/risk-k6-test/kafka/truststore_nonprod.jks, OriginalError: %!w(*errors.errorString=&{entry not found})
        at github.com/mostafa/xk6-kafka.(*Kafka).loadJKSFunction-fm (native)
        at file:///Users/jerry.liu/user/workspace_k6/risk-k6-test/kafka/loadjkstest.js:9:20(27)  hint="script exception"_

seems it is caused by that even it is expected to return the jks according to the comments below in the [code](https://github.com/mostafa/xk6-kafka/blob/912ba4d2cf4b725a0c2beb750f730d5fa0fca773/jks.go#L4), it still throws the exception.
// Server CA is loaded, so it will be returned, even though the client key is not loaded.
		// This is because one might not have enabled mutual TLS (mTLS).
I raised a PR 
https://github.com/mostafa/xk6-kafka/pull/224, could you help to see whether it could help?

**Comment by mostafa** [comment]
@luciowu-airwallex 

There is an implementation bug I found that I am going to fix very soon.

**Comment by mostafa** [comment]
@luciowu-airwallex @arpitaso 

Please test the changes in #229 and provide feedback.

**Comment by luciowu-airwallex** [comment]
@mostafa ,  thank you for your fixing. I got the same error with my script in https://github.com/mostafa/xk6-kafka/issues/222#issuecomment-1600612379
```
ERRO[0000] GoError: Failed to decode client's private key: ./truststore.jks, OriginalError: %!w(*errors.errorString=&{entry not found})
        at github.com/mostafa/xk6-kafka.(*Kafka).loadJKSFunction-fm (native)
        at file:///Users/lucio.wu/code/test/risk-k6-test/kafka_jks.js:8:20(27)  hint="script exception"
```

I changed the code locally and if I updated the https://github.com/mostafa/xk6-kafka/blob/fix-load-jks-implementation/jks.go#L144C13-L144C13
as the following code, it is fine at my side
```
if jks == nil && err != nil {
		log.Println("why there is error here")
		common.Throw(runtime, err)
	}
```
 because we don't want to throw an exception if the jks is returned when only server CA is loaded in https://github.com/mostafa/xk6-kafka/blob/fix-load-jks-implementation/jks.go#L84

Could you help double-check it? Thanks

**Comment by mostafa** [comment]
@luciowu-airwallex 

I updated that line. Can you please try again?

**Comment by mostafa** [comment]
I'll release a version soon.

**Comment by mostafa** [comment]
@luciowu-airwallex @arpitaso 

Released [v0.18.2](https://github.com/mostafa/xk6-kafka/releases/tag/v0.18.2) with the fix.

**Comment by luciowu-airwallex** [comment]
Thanks! @mostafa
